### PR TITLE
Deezerfull

### PR DESCRIPTION
**Pull request by deezerdl**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

This pull request adds full deezer download support, it supports track, album, artist and playlist urls.
The original youtube-dl deezer extractor only supports 30 second samples from playlist urls.

### CODE REVIEW

File: youtube_dl/extractor/deezer.py [hunk 1]
```diff
@@ -10,9 +10,53 @@
...
+    return "".join(chr(ord(h[i]) ^ ord(h[i+16]) ^ ord(key[i])) for i in range(16))
+
+
+def getformat(song):
+    """ return format id for a song """
+    if song["FILESIZE_MP3_320"]:
+        return 3
+    if song["FILESIZE_MP3_256"]:
+        return 5
+    return 1
```
Comment (dstftw) [inline youtube_dl/extractor/deezer.py]: This code seems to be taken from https://github.com/Hopfengetraenk/DeezerDownload/blob/master/DeezerDownload.py.
You are only allowed to use codes released to public domain so you have to ask its original author to release in under Unlicense.

File: youtube_dl/downloader/deezer.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+from Crypto.Cipher import Blowfish
```
Comment (dstftw) [inline youtube_dl/downloader/deezer.py]: This may not be available. Graceful message should be printed in this case.

**Comment by dstftw** [comment]
1. Read new extractor tutorial and coding conventions and fix your code.
2. Check code with flake8.
3. Does not work at all:
```
PS C:\Dev\youtube-dl\master> py -3.5 .\youtube_dl\__main__.py http://www.deezer.com/playlist/176747451 -F
[DeezerPlaylist] 176747451: Downloading webpage
Traceback (most recent call last):
  File ".\youtube_dl\__main__.py", line 19, in <module>
    youtube_dl.main()
  File "C:\Dev\youtube-dl\master\youtube_dl\__init__.py", line 465, in main
    _real_main(argv)
  File "C:\Dev\youtube-dl\master\youtube_dl\__init__.py", line 455, in _real_main
    retcode = ydl.download(all_urls)
  File "C:\Dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 1896, in download
    url, force_generic_extractor=self.params.get('force_generic_extractor', False))
  File "C:\Dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 760, in extract_info
    ie_result = ie.extract(url)
  File "C:\Dev\youtube-dl\master\youtube_dl\extractor\common.py", line 433, in extract
    ie_result = self._real_extract(url)
  File "C:\Dev\youtube-dl\master\youtube_dl\extractor\deezer.py", line 123, in _real_extract
    urlkey = calcurlkey(int(s["SNG_ID"]), str(s["MD5_ORIGIN"]), int(s["MEDIA_VERSION"]), getformat(s))
  File "C:\Dev\youtube-dl\master\youtube_dl\extractor\deezer.py", line 33, in calcurlkey
    data = b'\xa4'.join([md5hex(data), data])+b'\xa4'
TypeError: sequence item 0: expected a bytes-like object, str found
```

File: youtube_dl/extractor/deezer.py (hunk 1 above)
Comment (deezerdl) [inline youtube_dl/extractor/deezer.py]: my original code is on http://haklab4ulibiyeix.onion/revres/DeezerDownloader/  ( use torbrowser to view )
or if you don't have torbrowser, you can use a tor2web gateway like: https://haklab4ulibiyeix.onion.link/revres/DeezerDownloader/

I originally hosted my code here on github, until my repo was taken down with a DMCA request.
Lots of copies of my original code still exist on github.

File: youtube_dl/downloader/deezer.py (hunk 2 above)
Comment (deezerdl) [inline youtube_dl/downloader/deezer.py]: i now check at the start of the extractor if pycrypto is installed.

**Comment by deezerdl** [comment]
my recent commits fix the python3 issues, and fixed code layout issues reported by flake8.

**Comment by jaimehrubiks** [comment]
is this the real reversing-research?

**Comment by deezerdl** [comment]
i will add a blowfish.py in the same way as there is already aes.py.

**Comment by deezerdl** [comment]
I removed the pycrypto dependency.

**Comment by deezerdl** [comment]
The travis build has two more errors:
-- python 2.6
  File "/home/travis/build/rg3/youtube-dl/youtube_dl/blowfish.py", line 295, in blowfish_cbc_decrypt
    ivl, ivr = struct.unpack(">LL", iv)
TypeError: Struct() argument 1 must be string, not unicode


-- python 3.2
  File "/home/travis/build/rg3/youtube-dl/test/test_blowfish.py", line 74, in test_cbc1
    key = a2b_hex("0123456789abcdeff0e1d2c3b4a59687")
TypeError: 'str' does not support the buffer interface


And then all the `YTDL_TEST_SET=download` builds fail, but not due to my code.

**Comment by deezerdl** [comment]
Now all  `YTDL_TEST_SET=core`   tests succeed.

**Comment by deezerdl** [comment]
there is still an issue with the thumbnails not set.  the original code only worked for playlists. i will have to change it to work with artist, album, track, and playlist.

**Comment by Kabouik** [comment]
Just curious and excited about Deezer support, would you have an update to share on this @deezerdl? Thanks for your work on this plugin.

**Comment by deezerdl** [comment]
I have been waiting for this to be merged for quite a while. I guess @dstftw is quite busy.
Travis, the CI tool, complains about half the tests it did run, those fail on all branches.

**Comment by jopasserat** [comment]
There is an edge case for playlists with more than 40 entries, only the first 40 ones are fetched.
I guess Deezer uses some form of lazy loading on the page, kind of like on Instagram, and you'd have to force a scroll down with something like Selenium?

Example of such a playlist: http://www.deezer.com/playlist/1227352431

**Comment by kenavera** [comment]
For some reason my deezer url's for me are of the form ` http://www.deezer.com/us/album/XXX ` or ` http://www.deezer.com/us/playlist/XXX` and they do not download, whilst when I remove the /us/ it does. If I go to deezer I'm always redirected to the /us/ version. Is something you can add to this pr?

**Comment by Notmarrco** [comment]
It seems that deezer released a new version that is no more compatible with this PR. :disappointed:

**Comment by deezerdl** [comment]
The crypto has not changed, but the MD5_ORIGIN key is no longer present when not logged in.
I would have to add some code to authenticate to deezer.

**Comment by deezerdl** [comment]
I have created another version of my tool, independent of youtube-dl, see http://haklab4ulibiyeix.onion/revres/DeezerDownloader

It allows you to decode a browser developer mode capture file, and extract audio files from it.

**Comment by woeterman94** [comment]
any update?

**Comment by Electrojones** [comment]
Any updates?

File: youtube_dl/extractor/deezer.py [hunk 3]
```diff
@@ -8,84 +8,192 @@
...
+    if len(data) % 16:
+        data += b'\x00' * (16 - len(data) % 16)
+    enc = aes_ecb_encrypt(bytes_to_intlist(data), bytes_to_intlist(b"jo6aey6haid2Teih"))
+    return b2a_hex(intlist_to_bytes(enc)).decode('utf-8')
+
+
+def calcblowfishkey(songid):
+    """ Calculate the Blowfish decrypt key for a given songid """
+    h = md5hex(str(songid).encode('ascii'))
+    key = b"g4el58wc0zvf9na1"
```
Comment (gamer191) [inline youtube_dl/extractor/deezer.py:40]: How did you obtain this key?

Comment (LucBerge) [reply youtube_dl/extractor/deezer.py:40]: It must be available somewhere in the browser but I could not find it. Installing a tool like [Firebug ](https://getfirebug.com/)is a good start to find it.

The @deezerdl user [has been created for this pull request](https://github.com/deezerdl?tab=overview&from=2017-12-01&to=2017-12-31), he will probably never answer...
